New feature: JSON file input

bashir2 opened the pull request:
<!-- This is based on openmrs-cord PR template. -->
## Description of what I changed
This PR adds the option for reading FHIR resources from JSON files (each file is expected to be one `Bundle`).

The main motivation for this was faster conversion of Synthea datasets but this is also the first
step in the direction of consuming JSON files produced by the bulk `$export` API of FHIR.

Please note only the second commit belongs to this PR; the first commit is from PR #283.
<!--- Describe your changes in detail -->
<!--- It can simply be your commit message, which you must have -->
<!--- If your PR is related to an issue , please mention it here. -->
<!--- It is generally a good practice to first file an issue with enough
  context and reference it in the PR, but if you don't have that, please remove
  this section. -->

## E2E test
<!-- There are different scenarios for using the tools in this repo; please 
  help your reviewers by describing how you have e2e tested your change. -->
TESTED:
I tested reading from files with both Parquet output and DB (jsonb); here is the Parquet example:

```
$ java -cp ./pipelines/batch/target/batch-bundled-0.1.0-SNAPSHOT.jar org.openmrs.analytics.FhirEtl \
  --outputParquetPath=tmp/OUT_768_patients_fixed-ref/ --resourceList="Patient,Encounter,Observation" \
  --runner=FlinkRunner --maxParallelism=10 --parallelism=10 \
  --sourceJsonFilePattern=[PATH]/synthea-hiv/768_patients/*
```

## Checklist: I completed these to help reviewers :)
<!--- Put an `x` in the box if you did the task -->
<!--- If you forgot a task please follow the instructions below -->
- [x] I have read and will follow the [review process](https://github.com/GoogleCloudPlatform/openmrs-fhir-analytics/blob/master/doc/review_process.md).
- [x] My IDE is configured to follow the [**code style**](https://wiki.openmrs.org/display/docs/Java+Conventions) of this project.
- [x] I am familiar with Google Style Guides for the language I have coded in.
- [x] I have **added tests** to cover my changes. (If you refactored existing code that was well tested you do not have to add tests)
- [x] I ran `mvn clean package` right before creating this pull request and added all formatting changes to my commit.
- [x] All new and existing **tests passed**.
- [x] My pull request is **based on the latest changes** of the master branch.

omarismail94 commented on pipelines/common/src/main/java/org/openmrs/analytics/ParquetUtil.java:
```diff
@@ -222,27 +228,44 @@ public List<GenericRecord> generateRecords(Bundle bundle) {
 		org.hl7.fhir.dstu3.model.Resource r3Resource = org.hl7.fhir.convertors.VersionConvertor_30_40
 		        .convertResource(resource, true);
+		if (r3Resource == null) {
+			log.error("R3 conversion failed; ignoring resource with type " + resource.getResourceType().name());
```
Are we OK with returning  `null` if `r3Resource` is `null`. I get why we return `null` for the `r3Resource.getResourceType()` case, but not this

bashir2 replied:
The general problem is that the R4->R3 conversion can fail for various reasons (e.g., a new resource). In all these cases we should return `null` and handle it outside. These issues should go away once we fix #285 which I am planning to do soon.

omarismail94 commented on pipelines/batch/src/main/java/org/openmrs/analytics/ReadJsonFilesFn.java:
```diff
@@ -0,0 +1,92 @@
+		try {
+			IBaseResource resource = parser.parseResource(fileContent);
+			if (!"Bundle".equals(resource.fhirType())) {
+				log.error("The content of file " + file.getMetadata().toString() + " is not a Bundle; type: "
```
If it is better practice, switch to using `String.format` as opposed to string concatenation

omarismail94 commented on pipelines/batch/src/main/java/org/openmrs/analytics/ReadJsonFilesFn.java:
```diff
@@ -0,0 +1,92 @@
+			processBundle(bundle, resourceTypes);
+		}
+		catch (DataFormatException | ClassCastException e) {
+			log.error("Cannot parse content of file " + file.getMetadata().toString() + e);
```
ditto about `String.format`

bashir2 replied:
This one is very simple, i.e., it has only strings at the end, so I left it (there is always a trade-off).

omarismail94 reviewed (commented):
Thanks for this change Bashir! Just a few comments. Im curious to know what you think about using a cached version of the `FhirContext`

bashir2 commented on pipelines/batch/src/test/resources/bundle_relative_ref.json:
```diff
@@ -0,0 +1,135 @@
+      "mode": "include"
+    }
+  } ]
+}
```
Done.

bashir2 reviewed (commented):
Thanks for the review; PTAL.

bashir2 commented:
To be clear the only change since last review is [this commit](https://github.com/GoogleCloudPlatform/openmrs-fhir-analytics/pull/284/commits/45e87cc89dfbcfcaecc00e2923136a267785db85).

The reason for the many commits is that the remote branches (on GitHub) had extra changes due to master merges through GitHub UI which causes push conflicts. So I had to fetch those changes and push everything together.

bashir2 commented on pipelines/batch/src/main/java/org/openmrs/analytics/FetchSearchPageFn.java:
```diff
@@ -142,7 +147,8 @@ private static synchronized JdbcConnectionUtil getJdbcConnectionUtil(String jdbc
 	public void setup() throws SQLException, PropertyVetoException {
-		FhirContext fhirContext = FhirContext.forR4();
+		log.info("Starting setup for stage " + stageIdentifier);
+		fhirContext = FhirContext.forR4();
```
Sorry I missed this comment in the previous round (thanks to GitHub review tool!). The reason we actually create and keep the `fhirContext` in `setup` is because it is expensive and we only do it once per worker. Your suggestion is also a good one. But, interestingly, the `Cached`  version was not available in the HAPI version we were using. So I increased the HAPI version in `pom.xml` and switched to cached version.

The stale dependency issue should go away once we fix #243 (soon).

bashir2 replied:
I spoke too soon: Increasing the version of HAPI has some R3->R4 implications too. Those seem to be easily fixable too but I don't want to expand the scope of this PR much more. For now, I added a TODO to fix this once we move to a newer HAPI.

bashir2 replied:
Resolving this conversation and submitting this PR as I have a huge formatting change PR which I prefer to apply on top of this (instead of dealing with formatting issues of this again). I think I have addressed all comments but please feel free to comment more and I will address in a separate PR.
